File: iosApp/TechBlogDigest/FeedViewModel.swift
import Foundation
import Shared

enum FeedUiState {
    case loading
    case success([Post])
    case error(String)
}

/// UNTESTED — Kotlin `suspend fun`s are expected to import into Swift as `async throws`
/// (standard Kotlin/Native Objective-C interop since Swift 5.5), but the exact generated
/// signature — especially `getFeedForIos`'s `[String]`/`[Post]` bridging — needs
/// confirming against Xcode's generated header on a Mac.
@MainActor
final class FeedViewModel: ObservableObject {
    @Published var uiState: FeedUiState = .loading

    private let repository: FeedRepository
    private let prefsStore: UserPrefsStore

    init(repository: FeedRepository, prefsStore: UserPrefsStore) {
        self.repository = repository
        self.prefsStore = prefsStore
        Task { await load(forceRefresh: false) }
    }

    func refresh() {
        Task { await load(forceRefresh: true) }
    }

    private func load(forceRefresh: Bool) async {
        do {
            let tags = Array(prefsStore.interestTags)
            let posts = try await repository.getFeedForIos(interestTags: tags, forceRefresh: forceRefresh)
            uiState = .success(posts)
        } catch {
            uiState = .error(error.localizedDescription)
        }
    }
}

File: androidApp/src/main/kotlin/com/engblog/android/ui/feed/FeedViewModel.kt
package com.engblog.android.ui.feed

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.engblog.android.data.UserPrefsStore
import com.engblog.shared.model.Post
import com.engblog.shared.repository.FeedRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

sealed interface FeedUiState {
    data object Loading : FeedUiState
    data class Success(val posts: List<Post>, val isRefreshing: Boolean = false) : FeedUiState
    data class Error(val message: String) : FeedUiState
}

class FeedViewModel(
    private val repository: FeedRepository,
    private val prefsStore: UserPrefsStore,
) : ViewModel() {
    private val _uiState = MutableStateFlow<FeedUiState>(FeedUiState.Loading)
    val uiState: StateFlow<FeedUiState> = _uiState.asStateFlow()

    init {
        load(forceRefresh = false)
    }

    fun refresh() = load(forceRefresh = true)

    private fun load(forceRefresh: Boolean) {
        viewModelScope.launch {
            val current = _uiState.value
            if (current is FeedUiState.Success) {
                _uiState.value = current.copy(isRefreshing = true)
            }
            val interestTags = prefsStore.interestTags.first()
            runCatching { repository.getFeed(interestTags, forceRefresh) }
                .onSuccess { posts -> _uiState.value = FeedUiState.Success(posts) }
                .onFailure { error ->
                    _uiState.value = FeedUiState.Error(error.message ?: "Couldn't load the feed")
                }
        }
    }
}
